client/src/main/scala/org/http4s/client/oauth1/SignatureAlgorithm.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.client.oauth1
+
+import org.http4s.client.oauth1.SignatureAlgorithm.Names._
+import org.http4s.client.oauth1.ProtocolParameter.SignatureMethod
+
+import java.nio.charset.StandardCharsets.UTF_8
+import java.util.Base64
+import javax.crypto
+import javax.crypto.Mac
+import javax.crypto.spec.SecretKeySpec
+
+object SignatureAlgorithm {
+
+  private[oauth1] object Names {
+
+    // algorithms implemented so far
+    val `HMAC-SHA1` = "HMAC-SHA1"
+    val `HMAC-SHA256` = "HMAC-SHA256"
+    val `HMAC-SHA512` = "HMAC-SHA512"
+
+    // other known algorithms (not yet implemented)
+    val PLAINTEXT = "PLAINTEXT"
+    val `RSA-SHA1` = "RSA-SHA1"
+    val `RSA-SHA256` = "RSA-SHA256"
+    val `RSA-SHA512` = "RSA-SHA512"
+  }
+
+  private[this] val AllMethods = List(HmacSha1, HmacSha256, HmacSha512)
+
+  /**
+   * Map a [[SignatureMethod]] protocol parameter to a [[SignatureAlgorithm]] implementation
+   *
+   * @param method The signature method protocol parameter
+   * @return The implementation, or an [[IllegalArgumentException]] if none were found
+   */
+  def fromMethod(method: SignatureMethod): SignatureAlgorithm =
+    AllMethods.find(_.name == method.headerValue)
+    .getOrElse(throw new IllegalArgumentException(s"Unrecognized headerValue '${method.headerValue}', Valid options are: ${AllMethods.map(_.name)}"))
+
+}
+
+/**
+ * Implementations for Oauth1 signatures.
+ */
+sealed trait SignatureAlgorithm {
+
+  /**
+   * @return The signature method name per the oauth1.0 spec
+   */
+  def name: String
+
+  /**
+   * Apply the implementation's algorithm to the input
+   *
+   * @param input The input value
+   * @param secretKey The secret key
+   * @return The base64-encoded output
+   */
+  def generate(input: String, secretKey: String): String
+
+  private[oauth1] def generateHMAC(input: String, algorithm: String, secretKey: String): String = {
+
+    val mac: Mac = crypto.Mac.getInstance(algorithm)
```

Comment:
So, I've spent a few hours racking my brain on this one.  I attempted both 1 and 2 but the problem is binary compatibility.  F ripples up to the `genAuthHeader` which I can't change, so I'm left with the other option of duplicating functions all the way down and at that point, I felt like I had a huge frankenstein's monster.

If that were the case, I think I'd spare the users of this class my ugly hacks and just make a local oauth1 copy that has been switched to SHA256 :)

Reply:
I will see if I can do anything locally, but I'd rather unblock your enhancement.  It will only throw in exotic cases, using functionality people didn't have before, so I'm comfortable with it.